workerpool.go

```diff
@@ -97,32 +102,45 @@ func (wp *workerPool) clean(scratch *[]*workerChan) {
 
 	// Clean least recently used workers if they didn't serve connections
 	// for more than maxIdleWorkerDuration.
-	currentTime := time.Now()
+	criticalTime := time.Now().Add(-maxIdleWorkerDuration)
 
 	wp.lock.Lock()
 	ready := wp.ready
 	n := len(ready)
-	i := 0
-	for i < n && currentTime.Sub(ready[i].lastUseTime) > maxIdleWorkerDuration {
-		i++
-	}
-	*scratch = append((*scratch)[:0], ready[:i]...)
-	if i > 0 {
-		m := copy(ready, ready[i:])
-		for i = m; i < n; i++ {
-			ready[i] = nil
+
+	// Use binary-search algorithm to find out the index of the least recently worker which can be cleaned up.
+	i := func(l, r int) int {
+		var mid int
+		for l <= r {
+			mid = (l + r) / 2
+			if criticalTime.After(wp.ready[mid].lastUseTime) {
+				l = mid + 1
+			} else {
+				r = mid - 1
+			}
 		}
-		wp.ready = ready[:m]
+		return r
+	}(0, n-1)
```

Comment:
What is the reason you are using a closure here instead of just inlining the code?

Reply:
No any technical purpose for this, just for code structure, I can transform them into inline code.